### PR TITLE
Tree Widget Spacing Fix

### PR DESCRIPTION
**Pull request by joehenry9498**
## Changes

Created a ControlledTree example in Example Components Provider. Fixed spacing issues with nodes not lining up when some have expanders and some do not.

## Testing

Test in the component examples provider. Make sure nodes with and without expanders line up with one another.

### CODE REVIEW

**Comment by saskliutas** [comment]
I have looked at Tree in test-app and it looks like adding blank div with `10px` width only works because expansion toggle has `margin-right: 10px` applied from unrelated css selector.
https://github.com/iTwin/appui/blob/71ab57c9416bb598688d13f47bff3563489d30ac/test-apps/appui-test-app/appui-test-providers/src/ui/frontstages/ComponentExamples.scss#L75

This is how Tree looks if `margin-right` is not applied:
image

**Comment by joehenry9498** [comment]
That margin-right styling was unneeded for components provider. Took it out and used a blank div instead of width: 10px

**Review by raplemie — CHANGES_REQUESTED** [review]

File: ui/core-react/src/core-react/tree/Node.tsx [hunk 1]
```diff
@@ -141,7 +141,9 @@ export class TreeNode extends React.Component<TreeNodeProps> {
     ) : undefined;
 
     const toggle =
-      this.props.isLoading || this.props.isLeaf ? undefined : (
+      this.props.isLoading || this.props.isLeaf ? (
+        <div />
```
Comment (saskliutas) [inline ui/core-react/src/core-react/tree/Node.tsx:145]: Looks like this empty `div`  causes some alignment issues when loading spinner is shown.
image

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: ui/core-react/src/core-react/tree/Node.tsx (hunk 1 above)
Comment (joehenry9498) [reply ui/core-react/src/core-react/tree/Node.tsx]: Empty div should no longer appear when loading is occurring

File: test-apps/appui-test-app/appui-test-providers/src/ui/widgets/TreeWidget.tsx [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    parent?: TreeNodeItem,
+    pageOptions?: PageOptions
+  ): Promise<TreeNodeItem[]> {
+    return this.getNodesByParentId(parent ? parent.id : undefined, pageOptions);
+  }
+
+  public async getNodesCount(parent?: TreeNodeItem): Promise<number> {
+    return this.getNodesByParentId(parent ? parent.id : undefined).length;
+  }
+}
```
[the remark this answers is not in the document]
Comment (joehenry9498) [reply test-apps/appui-test-app/appui-test-providers/src/ui/widgets/TreeWidget.tsx]: Completed

**Comment by raplemie** [comment]
image
run `rush update --variant core-3x` so this gets fixed.

File: test-apps/appui-test-app/appui-test-providers/src/ui/providers/ComponentExamplesProvider.tsx [hunk 3]
```diff
@@ -1277,6 +1278,21 @@ export class ComponentExamplesProvider {
...
 
+  private static get treeSample(): ComponentExampleCategory {
+    return {
+      title: "Controlled Tree Widget",
+      examples: [
+        createComponentExample(
+          "Controlled Tree Widget",
+          undefined,
+          <div className="tree-widget-tree-container">
+            <TreeWidgetComponent />
```
[the remark this answers is not in the document]
Comment (joehenry9498) [reply test-apps/appui-test-app/appui-test-providers/src/ui/providers/ComponentExamplesProvider.tsx]: All good!

**Comment by joehenry9498** [comment]
Everything should be good now

**Review by raplemie — APPROVED** [review]
LGTM!

File: docs/storybook/src/components/TreeWidget.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  const modelSource = useTreeModelSource(dataProvider);
+
+  const nodeLoader = props.nodeLoader
+    ? props.nodeLoader
+    : useTreeNodeLoader(dataProvider, modelSource);
+
+  const eventHandler = props.eventsHandler
+    ? props.eventsHandler
+    : useTreeEventsHandler({ modelSource, nodeLoader });
+  const treeModel = props.model ? props.model : useTreeModel(modelSource);
```
[the remark this answers is not in the document]
Comment (joehenry9498) [reply docs/storybook/src/components/TreeWidget.stories.tsx]: Removed them

File: test-apps/appui-test-app/appui-test-providers/src/ui/widgets/TreeWidget.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+export function TreeWidgetComponent() {
+  const [dataProvider] = React.useState(
+    () => new SimpleTreeDataProvider(hierarchyTest)
+  );
+
+  const modelSource = useTreeModelSource(dataProvider);
+
+  const nodeLoader = useTreeNodeLoader(dataProvider, modelSource);
+
+  const eventHandler = useTreeEventsHandler({ modelSource, nodeLoader });
```
Comment (saskliutas) [inline test-apps/appui-test-app/appui-test-providers/src/ui/widgets/TreeWidget.tsx:79]: Params to `useTreeEventsHandler` need to be memoized to make sure new event handler is not created on each render.

```suggestion
  const eventHandler = useTreeEventsHandler(React.useMemo(() => ({ modelSource, nodeLoader }), [modelSource, nodeLoader]));
```

**Review by saskliutas — APPROVED** [review]

**Pull request merged**